components/service/glean/src/main/java/mozilla/components/service/glean/storages/PingStorageEngine.kt

```diff
@@ -0,0 +1,133 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean.storages
+
+import android.content.Context
+import android.support.annotation.VisibleForTesting
+import java.io.BufferedReader
+import java.io.BufferedWriter
+import java.io.File
+import java.io.FilenameFilter
+import java.io.FileNotFoundException
+import java.io.FileOutputStream
+import java.io.FileReader
+import java.io.IOException
+import java.io.OutputStreamWriter
+import java.util.regex.Matcher
+import java.util.regex.Pattern
+
+import mozilla.components.support.base.log.logger.Logger
+
+/**
+ * TelemetryStorage implementation that stores pings as files on disk.
+ */
+internal open class PingStorageEngine(context: Context) {
+
+    private val logger: Logger = Logger("glean/PingStorageEngine")
+
+    @VisibleForTesting
+    internal val storageDirectory: File = File(context.applicationInfo.dataDir, STORAGE_DIRECTORY)
+
+    companion object {
+        private const val FILE_PATTERN = "[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}"
+        private const val STORAGE_DIRECTORY = "glean/storage"
+    }
+
+    @Synchronized
+    fun store(pingType: String, pingPath: String, ping: String) {
+        val pingStorageDirectory = File(storageDirectory, pingType)
+        assertDirectory(pingStorageDirectory)
+
+        // Build the file path for the ping, using the UUID from the path for the file name
+        val pingFile = File(pingStorageDirectory, pingPath.substring(pingPath.lastIndexOf('/')))
```

Comment:
Maybe pass in the UUID and use it directly.  I feel like parsing the URL path might a bit brittle if the backend ever changes how they want paths to be handled.